wonkyconn/features/age_sex_prediction.py

```diff
@@ -0,0 +1,387 @@
+from __future__ import annotations
+
+import time
+import warnings
+from typing import Dict, List, Optional
+
+import numpy as np
+from joblib import parallel_backend
+from sklearn.decomposition import PCA
+from sklearn.impute import SimpleImputer
+from sklearn.linear_model import LogisticRegression, Ridge
+from sklearn.model_selection import (
+    ShuffleSplit,
+    StratifiedShuffleSplit,
+    cross_validate,
+)
+from sklearn.pipeline import Pipeline
+from sklearn.preprocessing import LabelEncoder, StandardScaler
+from sklearn.svm import SVC, SVR
+
+# ---------------------------------------------------------------------
+# Global caps
+# ---------------------------------------------------------------------
+FIXED_N_SPLITS = 100
+MAX_PCA = 100
+
+
+# ---------------------------------------------------------------------
+# RunReport for concise summary
+# ---------------------------------------------------------------------
+class RunReport:
+    def __init__(self):
+        self.sex_error: Optional[str] = None
+        self.age_error: Optional[str] = None
+        self.notes: Dict[str, str] = {}
+
+    def set_note(self, key: str, msg: str):
+        self.notes[key] = msg
+
+    def set_sex_error(self, msg: str):
+        self.sex_error = msg
+
+    def set_age_error(self, msg: str):
+        self.age_error = msg
+
+    def short_summary(self) -> str:
+        parts = []
+        if self.sex_error:
+            parts.append(f"sex prediction failed: {self.sex_error}")
+        if self.age_error:
+            parts.append(f"age prediction failed: {self.age_error}")
+        if not parts:
+            parts.append("both predictions succeeded")
+        return " | ".join(parts)
+
+
+# ---------------------------------------------------------------------
+# Vectorization helper
+# ---------------------------------------------------------------------
+def stack_upper_triangle(connectivity_matrices: List["ConnectivityMatrix"]) -> np.ndarray:
```

Comment:
A lot of this function can be replaced by `nilearn.connectome.sym_matrix_to_vec`, and you can avoid writing your own checks.

https://nilearn.github.io/dev/modules/generated/nilearn.connectome.sym_matrix_to_vec.html

Since you wrote zero unit tests, I suggest using existing tools as much as possible.